### ISSUE TITLE
Mining algorithm negotiation

### ISSUE THREAD

**Issue by xmrig** [xmrig/xmrig-proxy#168]
I have plan to extend stratum protocol to make possible algorithm negotiation between miner and pool/proxy. [Current working draft](https://github.com/xmrig/xmrig-proxy/blob/dev/doc/STRATUM_EXT.md#mining-algorithm-negotiation). Any feedback welcome.

**Comment by el00ruobuob** [comment]
For cn-heavy variant, i believe it should support it with only variant 0 (or -1) at the moment. It allows evolutions.

**Comment by el00ruobuob** [comment]
And why the proxy shouldn't send the variant? It could tell the miner which one to use. IMO it's up to the job requester to choose it.

**Comment by xmrig** [comment]
Your are right about cn-heavy, I changed docs to better clarify this moment.

Variant=-1 same as missing variant, works fine for Monero and Aeon, all miners will detect proper variant from block version, manual variant required only for coins where used non standard block header version, sad but other coins authors don't like to change version to 7 to make it work without variant.

Miner never request job from pool, pool always send job, it completely up to pool. In protocol has method  `getjob`, but miners not use it and no useful reason to use it, also I found this method is buggy.
Thank you.

**Comment by el00ruobuob** [comment]
I mean pool/proxy could impose variant when sending the job.

**Comment by mpkogli** [comment]
I agree that pool should be able to "suggest" the algorithm to the miner via the proxy. Perhaps this will require re-jiggering on the miner side, but would help with the current algo war.

**Comment by MoneroOcean** [comment]
Good proposal. I will support it in MO pool and MO xmr-node-proxy.

**Comment by mpkogli** [comment]
@NicNull Yup, but this would be handled on the miner side. Theoretically...

**Comment by aleqx** [comment]
I support this. Also, let's forget about pools for now (I also don't think they will ever support this extension). 

This extension is massively helpful as it is: many of us have multiple miners. It's much, much easier and time effective to change the settings only in the proxy without touching the miners. It's even more awesome that we don't have to stop the proxy either since the new xmrig-proxy versions autodetect config file changes and reload it. I would love to be able to switch between **any** coin by changing only the xmrig-proxy config (right now I can only switch variants).

The set of algos and variants in the xmrig-proxy should match those supported by xmrig, and the config for both should be identical so that folks can easily use them.

I don't think it matters too much how you do it in json-rpc, because I don't think any pool will ever implement this extension.

However, it woul dbe of interest to make the extension compatible with current imeplementations,so that xmrig-proxy can be used with other miners too, and not just xmrig (e.g. xmr-stak). I see you changed `params` from an array to an object ... that would break compatibility. Why not keep it as an array but add 3 values at the end, regardless of the length of `params:[]`. I'd make the last value in params always equal to "xmrig-proxy-ext.v1" (or some other identifier for the protocol extension), and then the next to last and next next to last values equal to the algo name and variant number.

Keep up the good work.

p.s. Personally I don't see the point in having two parameters `algo` and `variant` when there could be just one ... the Cryptonight landscape is getting out of hand.

**Comment by xmrig** [comment]
I don't change `params` from an array to an object, in stratum protocol always used an object. All changes compatible with exists miners including xmr-stak.

**Comment by aleqx** [comment]
All the `params` I see on https://github.com/xmrig/xmrig-proxy/blob/dev/doc/STRATUM_EXT.md#mining-algorithm-negotiation are objects `params:{}` instead of arrays `params:[]` ... but maybe I'm missing something.

**Comment by xmrig** [comment]
I added `cryptonight-ipbc` to specification.
@aleqx  `params:{}` it correct and current protocol, all extensions is backward compatible with all exists miners.

**Comment by armin884** [comment]
I'm mining haven (cn-heavy) with xmrig-proxy/2.5.3 and miner is XMRig/2.6.0-beta3 . How do i get auto algorithm negotiation working? If i don't specify algo on miner with -a cryptonight-heavy it always uses cryptonight and i get Low difficulty share error

**Comment by aleqx** [comment]
@xmrig does master for xmrig-proxy and xmrig-nvidia now support changing both algo and variant? Concretely, if I connect xmrig miners to xmrig-proxy, and I change both the algo and variant in the xmrig-proxy config, will those be pushed to the miner as well (i.e. I don't have to restart miners with a different config)?

What about pushing different `threads`, `blocks`, `bfactor`, `bsleep` parameters from xmrig-proxy to xmrig to avoid restarting xmrig with different configs?

**Comment by mpkogli** [comment]
@xmrig I'm wondering the same thing as @aleqx. How do GPU configurations figure into the mix? Does any miner support automatically changing GPU threads/blocks/intensity without restarting the miner? Thanks!

**Comment by unsocial-bleach** [comment]
Is there any plan to finish the last item on this list:

>  Multi algorithm mode in proxy (router).

This feature would be very helpful for me, and would allow us to close this issue.
